### PR TITLE
feat(experimental): add `programNotifications` websocket method

### PR DESCRIPTION
**Pull request by buffalojoec**

### CODE REVIEW

**Comment by buffalojoec** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1643** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1643" target="_blank">Graphite</a> 
    * **PR #1644** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1644" target="_blank">Graphite</a> 
      * **PR #1645** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1645" target="_blank">Graphite</a> 
        * **PR #1646** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1646" target="_blank">Graphite</a> 
          * **PR #1647** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1647" target="_blank">Graphite</a> 
            * **PR #1648** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1648" target="_blank">Graphite</a> 
              * **PR #1649** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1649" target="_blank">Graphite</a> 
                * **PR #1650** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1650" target="_blank">Graphite</a> 
                  * **PR #1651** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1651" target="_blank">Graphite</a>  👈
                    * **PR #1652** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1652" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1651?utm_source=stack-comment).

File: packages/rpc-core/src/rpc-subscriptions/program-notifications.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+} from '../rpc-methods/common';
+
+type ProgramNotificationsMemcmpFilter = Readonly<{
+    offset: U64UnsafeBeyond2Pow53Minus1;
+    bytes: string;
+    encoding: 'base58' | 'base64';
+}>;
+
+type ProgramNotificationsDatasizeFilter = Readonly<{
+    dataSize: U64UnsafeBeyond2Pow53Minus1;
```
Comment (steveluscher) [inline packages/rpc-core/src/rpc-subscriptions/program-notifications.ts:22]: Kinda crazy that this is actually a `u64`.

File: packages/rpc-core/src/rpc-subscriptions/program-notifications.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    AccountInfoWithBase64EncodedZStdCompressedData,
+    AccountInfoWithJsonData,
+    Commitment,
+    RpcResponse,
+    U64UnsafeBeyond2Pow53Minus1,
+} from '../rpc-methods/common';
+
+type ProgramNotificationsMemcmpFilter = Readonly<{
+    offset: U64UnsafeBeyond2Pow53Minus1;
+    bytes: string;
```
Comment (steveluscher) [inline packages/rpc-core/src/rpc-subscriptions/program-notifications.ts:17]: This type needs to be parametric on `encoding`. When it's `base58` it should be `Base58EncodedBytes` and when `base64` it should be `Base64EncodedBytes`.

You can probably either do this by having a union type that's discriminated on `encoding` and then doing some nonsense in the method signatures, but you might also be able to do it by adding a type parameter to `ProgramNotificationMemcmpFilter` and then doing some `bytes: TEncoding extends 'base64' ? Base64EncodedBytes : Base58EncodedBytes` trickery.

**Review by steveluscher — CHANGES_REQUESTED** [review]
[closetreadmillcat](https://app.graphite.dev/macros/solana-labs)

File: packages/rpc-core/src/rpc-subscriptions/program-notifications.ts (hunk 2 above)
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-subscriptions/program-notifications.ts:17]: I can do this, but we are not doing this lifting for `getProgramAccounts` HTTP methods

File: packages/rpc-core/src/rpc-subscriptions/program-notifications.ts (hunk 1 above)
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-subscriptions/program-notifications.ts:22]: Def overkill 😂

File: packages/rpc-core/src/rpc-subscriptions/program-notifications.ts (hunk 2 above)
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-subscriptions/program-notifications.ts:17]: It's done ✅

Comment (buffalojoec) [reply packages/rpc-core/src/rpc-subscriptions/program-notifications.ts:17]: It's done ✅

**Pull request merged**